### PR TITLE
allow hiercluster to work with numeric dictionary term data

### PR DESCRIPTION
**Pull request by congyu-lu**
## Description

please test will sjpp branch: add_numericDicTerm_cluster_all-pharmacotyping

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

File: client/plots/matrix/hierCluster.renderers.js [hunk 1]
```diff
@@ -10,6 +10,11 @@ export function maySetSandboxHeader() {
 		case TermTypes.METABOLITE_INTENSITY:
 			this.dom.header.text('Metabolite Intensity Clustering')
 			break
+		case termTypes.NUMERIC_DICTIONARY_TERM:
```
Comment (siosonel) [inline client/plots/matrix/hierCluster.renderers.js:13]: this should be uppercase `TermTypes`, I get an error when I try to load `Drug Sensitivity` chart.

**Comment by karishma-gangwani** [comment]
do we want to enable loading a default matrix view with all the normalized LC50 values?

**Comment by congyu-lu** [comment]
> do we want to enable loading a default matrix view with all or some of the normalized LC50 values?

yes can do it in another PR

File: client/plots/matrix/hierCluster.renderers.js (hunk 1 above)
Comment (congyu-lu) [reply client/plots/matrix/hierCluster.renderers.js]: thanks just fixed it

File: client/mass/charts.js [hunk 2]
```diff
@@ -280,6 +280,43 @@ function getChartTypeList(self, state) {
 				action.config.dataType = TermTypes.METABOLITE_INTENSITY
 				action.config.termgroups = [{ name: 'Metabolite Intensity Cluster', lst: twlst, type: 'hierCluster' }]
 			}
+		},
+		{
+			//use the app name defined in dataset file
+			label: state.termdbConfig.numericDictTermCluster?.appName || 'Numeric Dictionary Term cluster',
+			chartType: 'numericDictTermCluster',
+			clickTo: self.showTree_selectlst,
+			usecase: { target: 'numericDictTermCluster', detail: { exclude: ['Demographics', 'Age'] } },
```
Comment (xzhou82) [inline client/mass/charts.js:289]: you cannot hardcode the term id!!!!

File: client/plots/matrix/hierCluster.config.js [hunk 3]
```diff
@@ -63,10 +63,10 @@ export async function getPlotConfig(opts = {}, app) {
 				} else {
 					throw `term type missing and cannot be assigned by dataType`
 				}
-			} else if (!['geneExpression', 'metaboliteIntensity', 'float'].includes(tw.term.type)) {
+			} else if (!['geneExpression', 'metaboliteIntensity', 'float', 'integer'].includes(tw.term.type)) {
```
Comment (xzhou82) [inline client/plots/matrix/hierCluster.config.js:66]: should replace with `numericTypes` from shared/terms.js, use a single source

File: client/plots/matrix/hierCluster.config.js [hunk 4]
```diff
@@ -63,10 +63,10 @@ export async function getPlotConfig(opts = {}, app) {
...
 				// May add other term type in hierCluster
 				throw 'term type not supported in hierCluster'
```
Comment (xzhou82) [inline client/plots/matrix/hierCluster.config.js:68]: use specific err message `term type is not numeric`

File: server/routes/termdb.cluster.ts [hunk 5]
```diff
@@ -77,7 +83,9 @@ async function getResult(q: TermdbClusterRequest, ds: any) {
 		_q.forClusteringAnalysis = true
 	}
 
-	const { term2sample2value, byTermId, bySampleId } = await ds.queries[q.dataType].get(_q) // too strong assumption on queries[dt], may not work for single cell
+	const { term2sample2value, byTermId, bySampleId } =
```
Comment (xzhou82) [inline server/routes/termdb.cluster.ts:86]: please change to if-else. easier to expand later
if(q.dataType='numericDictterm') {
  getNumericDictTermAnnotation()
} else {
  await ds.queries[q.dataType]?.get(_q)
}

File: shared/types/src/routes/termdb.cluster.ts [hunk 6]
```diff
@@ -42,7 +43,17 @@ export type TermdbClusterRequestMetabolite = TermdbClusterRequestBase & {
 	terms: MetaboliteIntensityTerm[]
 }
 
-export type TermdbClusterRequest = TermdbClusterRequestGeneExpression | TermdbClusterRequestMetabolite
+export type TermdbClusterRequestNumericDictTerm = TermdbClusterRequestBase & {
+	/** Data type */
+	dataType: 'numericDictTerm'
+	/** List of terms */
+	terms: NumericDictTerm[]
```
Comment (xzhou82) [inline shared/types/src/routes/termdb.cluster.ts:50]: are they term objects or tw?

File: shared/types/src/dataset.ts [hunk 7]
```diff
@@ -915,6 +915,10 @@ type MatrixSettings = {
 	cnvUnit?: 'log2ratio' | 'segmedian'
 }
 
+type NumericDictTermClusterSettings = {
+	termGroupName?: string
```
Comment (xzhou82) [inline shared/types/src/dataset.ts:919]: explain purpose

File: shared/types/src/dataset.ts [hunk 8]
```diff
@@ -927,6 +931,14 @@ type Matrix = {
 	legendValueFilter?: any
 }
 
+type NumericDictTermCluster = {
+	/** alternative name, e.g. the plot is called "drug sensitivity" in ALL-pharmacotyping; by default it's called "Numeric Dictionary Term cluster" */
+	appName?: string
+	/** default settings for numericDictTermCluster plot */
+	settings?: NumericDictTermClusterSettings
+	excludes?: string[]
```
Comment (xzhou82) [inline shared/types/src/dataset.ts:939]: explain purpose

**Comment by karishma-gangwani** [comment]
another question. in ALL-pharmacotyping ds since normalized values are between 0 to 1, why is z-score calculation needed for clustering? it's already normalized data and should show the color scale for LC50 from 0 to 1 and not z-score. this is not gene expression data although they have that data as well.

**Comment by karishma-gangwani** [comment]
separate note. @robinpaul85 for the euclidean and canberra clustering are there proper definitions we can provide on the UI? similar to the maximum and manhattan to make it more meaningful?

**Comment by xzhou82** [comment]
exactly. there's no need to transform z-score here. this will be addressed by a separate fix https://github.com/stjude/proteinpaint/issues/2310

**Comment by robinpaul85** [comment]
> separate note. @robinpaul85 for the euclidean and canberra distance methods are there proper definitions we can provide on the UI when hovered over? similar to the maximum and manhattan to make it more meaningful?

I added what was mentioned in the hclust function documentation. For euclidean distance it should be easy. However, for canberra distance the explanation might be too long to mention. This is a good place to read more about it https://medium.com/@eskandar.sahel/exploring-common-distance-measures-for-machine-learning-and-data-science-a-comparative-analysis-ea0216c93ba3

**Comment by karishma-gangwani** [comment]
> > separate note. @robinpaul85 for the euclidean and canberra distance methods are there proper definitions we can provide on the UI when hovered over? similar to the maximum and manhattan to make it more meaningful?
> 
> I added what was mentioned in the hclust function documentation. For euclidean distance it should be easy. However, for canberra distance the explanation might be too long to mention. This is a good place to read more about it https://medium.com/@eskandar.sahel/exploring-common-distance-measures-for-machine-learning-and-data-science-a-comparative-analysis-ea0216c93ba3

 we can make it concise but it should inform the user about why to pick a certain method over the other.

**Comment by xzhou82** [comment]
might link to https://en.wikipedia.org/wiki/Canberra_distance to explain canberra distance

File: shared/utils/src/terms.js [hunk 9]
```diff
@@ -36,7 +36,8 @@ export const TermTypes = {
 	SAMPLELST: 'samplelst',
 	METABOLITE_INTENSITY: 'metaboliteIntensity',
 	SINGLECELL_GENE_EXPRESSION: 'singleCellGeneExpression',
-	SINGLECELL_CELLTYPE: 'singleCellCellType'
+	SINGLECELL_CELLTYPE: 'singleCellCellType',
+	NUMERIC_DICTIONARY_TERM: 'numericDictTerm'
```
Comment (xzhou82) [inline shared/utils/src/terms.js:40]: this shouldn't be a term type

Comment (congyu-lu) [reply shared/utils/src/terms.js]: thanks just moved it out

**Review by xzhou82 — APPROVED** [review]

**Pull request merged**